lib/identity_cache/query_api.rb

```diff
@@ -399,10 +405,12 @@ def fetch_recursively_cached_association(ivar_name, association_name) # :nodoc:
       if IdentityCache.should_use_cache?
         ivar_full_name = :"@#{ivar_name}"
 
-        unless ivar_value = instance_variable_get(ivar_full_name)
-          ivar_value = IdentityCache.map_cached_nil_for(send(association_name))
-          instance_variable_set(ivar_full_name, ivar_value)
-        end
+        ivar_value = if instance_variable_defined?(ivar_full_name)
+                       instance_variable_get(ivar_full_name)
+                     else
+                       value = IdentityCache.map_cached_nil_for(send(association_name))
```

Comment:
`IdentityCache.map_cached_nil_for` was to distinguish between `nil` and unloaded.  I think we should instead use the instance variable defined check to distinguish between these two states.  Since that isn't an isolated change, it should probably be handled in a separate PR.